README.md
# BERandomizer
Personal project aimed to randomize everything possible within Minecraft Bedrock Edition
## Current Features
- Automatic updates (can be disabled in a future version)
- Fully randomize all available recipes in BE 1.12, 1.13, and 1.14
    - Should support future recipes as they're added with little or no modification
- Allows for multiple datasets
    - Can store 1.12, 1.13, and 1.14 behavior data
    - Lets you choose which one to randomize
    - Supports randomizing custom data as well
        - Just drop it in it's own folder
        - data/DATASETNAME/(data contents)
            - ex. data/Vanilla-1.12/stuff
- Generates based on seed
    - Input same seed, get the same results.
    - Accepts strings or numbers, go wild.
        - Y'all know how this works. :p
- Settings
    - Toggle automatic updates
    - Toggle recipe randomization by type
## Planned Features
- Store both resource and behavior data
	- Needed to start randomizing blocks
- Add data values to documentation
- Automatically download data if nothing is found
- More error handling
    - Verify data has manifest.json
        - Possibly hard code manifest creation?
- Code optimization
    - Definitely needs work
- User-friendliness improvements
- More randomness
    - loot tables
    - entities?
    - blocks?
    - randomly generated custom blocks/items/entities?
        - this one will be fun!

## Usage
In order to use this, make sure y'all have Python installed, once that is done, download this repo, extract if necessary, and click on Run.bat to start the script.
Follow the prompts on screen to get started.
Remember to load your datasets

For Automatic Updates to work, you'll need to `git clone` the repository and run it from there. If this is not done and you just download the .zip, you will not recieve Automatic Updates
Any questions on anything, contact me on Discord with Fire#4224

Changelog.txt
10/15/2019 - Day One
- Initial code created, randomized recipes and archived into .mcpack

10/16/2019
- Split initial code into functions, improved readability
- Now can load multiple datasets
- Added seeded generation

10/17/2019
- Added automatic updates for those who 'git clone' the repository
    - Will be toggleable later

10/18/2019
- Improved user friendliness
- Added command system
    - Will likely be reworked later into its own module
- Reworked dirs variables
- Reworked a few dirs functions
- No longer hard coding directories
- Now erases temp data on boot if found

10/19/2019
- Optimized recipe randomization functions
- Fixed bug with recipe loading for 1.13
    - Added new recipe type: minecraft:recipe_brewing_container
- Added settings command
    - Toggle automatic updates
    - Add/Remove items to recipe blacklist

start.py
import json, os, random, time, subprocess, sys
from modules import dirs, recipes, autoupdate, settings
from modules.debug import debug

#dirsetup - making sure nothing breaks
dirs.cleanup(dirs.tempFolder) #in case of crash in last run
dirs.makedir(dirs.home+dirs.dataFolder) #makes data directory if it can't find it
subdirs = [o for o in os.listdir(dirs.dataDir) if os.path.isdir(os.path.join(dirs.dataDir,o))]

#controls the header
def header():
    '''Draws the header at the top of the screen'''
    subprocess.run("cls",shell=True)
    print("-----BERandomizer-----")
    print("Made by @InValidFire")
    print("Version: "+autoupdate.currentString)
    print("Automatic Updates: "+str(settings.settings['autoUpdate']))
    print("Seed: "+dirs.seed)
    print("Successfully loaded "+str(len(subdirs))+" dataset(s)")
    print("Selected dataset: "+str(dirs.datasetFolder.replace("\\","")))
    print("WARNING: Program is in active development, no garauntee of functionality is granted")
    print("----------------------")

#seed initialize
dirs.setSeed(int(time.time()))

#update check
settings.loadSettings(dirs.dataDir+"\\settings.json")
autoupdate.update()
header()

#directory handling - move to its own file
if(len(subdirs)==0):
    print("No data found in the"+dirs.dataFolder+" directory.\nExtract Behavior Pack data in a sub-folder of "+dirs.dataFolder+" to continue.")
    print("\nEx. "+dirs.dataFolder+"\\DATASETNAME\\(data contents)")
    exit()
if(os.path.exists(dirs.dataDir+"\\manifest.json")):
    print("Found a manifest.json in the "+dirs.dataFolder+" directory. Ensure you extracted into a subfolder of "+dirs.dataFolder+", not directly there.")
    print("\nEx. "+dirs.dataFolder+"\\DATASETNAME\\(data contents)")
    exit()
debug("Subdirs: "+str(subdirs))
datas = {}
if (len(subdirs)>1):
    i = 1
    for data in subdirs:
        datas[i] = data
        i = i+1
    print("Multiple datasets found:")
    for data in datas:
        print("\t"+str(data)+". "+str(datas[data]))
    debug("Keys: "+str(datas.keys()))
    dirs.setdatasetFolder(datas[int(input("Select a number to load dataset: "))])
    subprocess.run("cls",shell=True)
    debug(dirs.datasetFolder)
elif(len(subdirs)==1):
    dirs.setdatasetFolder(subdirs[0])
    debug(dirs.datasetFolder)

#welcome screen
header()
print("'help' to view commands\n")

#command system - move to its own file
while(True):
    command = input("Enter a command to continue: ")
    if(command.lower()=="help"):
        print("Available commands:\n\tseed - change the used seed\n\trandomize - start randomization process\n\tsettings - change settings for the randomizer\n\texit - close program")
    elif(command.lower()=="seed"):
        seed = input("Enter seed to generate: ")
        if(len(seed)<=0): #if seed is blank, use system time.
            seed = int(time.time())
            dirs.setSeed(seed)
        dirs.setSeed(seed)
        random.seed(seed)
        header()
    elif(command.lower()=="settings"):
        header()
        for item in settings.settings:
            print(item+" = "+str(settings.settings[item]))
        setting = input("Which setting would you like to change? ")
        if(setting=='autoUpdate'):
            print("Toggling automatic updates!")
            if(settings.settings[setting]==True):
                settings.setUpdate(False)
            else:
                settings.setUpdate(True)
            header()
        if(setting=='recipeBlacklist'):
            print("Types currently in blacklist: ")
            if(len(settings.settings['recipeBlacklist'])==0):
                print("None")
            for item in settings.settings['recipeBlacklist']:
                print(str(item))
            print("\nPossible types to add: ")
            for item in recipes.loadlist:
                if item not in settings.settings['recipeBlacklist']:
                    print(str(item))
            mode = input("Would you like to add or remove a recipe type? ")
            if(mode.lower()=="add"):
                rtype = input("Enter the type name: ")
                if(rtype not in settings.settings['recipeBlacklist']):
                    if(rtype in recipes.loadlist):
                        settings.editrecipeFilter("add",rtype)
                    else:
                        print("Could not find recipe type.")
                else:
                    print("Recipe type already blacklisted")
            if(mode.lower()=="remove"):
                rtype = input("Enter the type name: ")
                if(rtype in recipes.loadlist):
                    if(rtype in settings.settings['recipeBlacklist']):
                        settings.editrecipeFilter("remove",rtype)
                    else:
                        print("Recipe type is not blacklisted")
                else:
                    print("Could not find recipe type.")
        settings.writeSettings(dirs.dataDir+"\\settings.json")
    elif(command.lower()=="randomize"): #starts recipe randomizing -> modules.recipes
        header()
        print("Found "+str(dirs.countDir(dirs.dataDir+"\\"+dirs.datasetFolder+"\\recipes"))+" recipes")
        print("Randomization Options")
        print("\t1. recipes")
        command = input("Select numbers to randomize: ")
        if("1" in command):
            header()
            recipes.recipestart()
        dirs.package()
        dirs.cleanup(dirs.tempFolder)
        sys.exit()
    elif(command.lower()=="exit" or command.lower()=="quit"):
        sys.exit()
    else:
        print("Command not found, type 'help' for a list of commands.")

modules/dirs.py
import os, shutil, json, uuid, random
from modules.debug import debug
home = os.getcwd()
tempFolder = "\\tmp"
tempDir = home+tempFolder
dataFolder = "\\data"
dataDir = home+dataFolder
datasetFolder = "Not loaded"
# below is needed to stop compiler warnings, not needed for code functionality
seed = None
randomized = None
vanilla = None

def setSeed(value):
    '''Sets the seed used to randomize'''
    global seed
    seed = str(value)
    random.seed(seed) #oops

def setdatasetFolder(name):
    '''Set the dataset folder to use'''
    global datasetFolder
    global randomized
    global vanilla
    datasetFolder="\\"+name
    debug("updated datasetFolder to "+datasetFolder)
    randomized = tempDir+datasetFolder+"\\randomized.txt"
    vanilla = tempDir+datasetFolder+"\\vanilla.txt"

def copydir(source,dest):
    '''Copy directory'''
    print("Copying "+source+" to "+dest)
    shutil.copytree(source,dest)

def makedir(path):
    '''Make a directory'''
    if(not os.path.exists(path)):
        os.makedirs(path)

def deldir(path):
    '''Delete a directory'''
    if(os.path.exists(home+path)):
        shutil.rmtree(home+path)

def copy(source,dest):
    '''Copy a file'''
    shutil.copy(source,dest)

def countDir(path):
    '''Count items in directory'''
    i = 0
    for item in os.listdir(path):
        i = i+1
    return i

def archive(source,dest):
    '''Archive directory'''
    shutil.make_archive(dest,"zip",source)

def package():
    '''Package into .mcpack'''
    print("Copying manifest")
    copy(home+dataFolder+datasetFolder+"\\manifest.json",tempDir+datasetFolder)
    print("Modifying manifest")
    with open(tempDir+datasetFolder+"\\manifest.json","r") as json_file:
        jsondata = json.load(json_file)
        json_file.close()
    with open(tempDir+datasetFolder+"\\manifest.json","w") as json_file:
        jsondata['header']['uuid'] = str(uuid.uuid4())
        jsondata['header']['name'] = "BERandomizer - Seed: "+str(seed)
        jsondata['header']['description'] = "Made by @InValidFire"
        jsondata['modules'][0]['uuid'] = str(uuid.uuid4())
        jsondata['modules'][0]['description'] = "Made by @InValidFire"
        jsondata['dependencies'] = []
        json.dump(jsondata,json_file,indent=4)
        json_file.close()
    print("Archiving into .mcpack")
    archive(tempDir+datasetFolder,home+"\\BERandomizer - "+str(seed)+" - "+datasetFolder.replace("\\",""))
    os.rename(home+"\\BERandomizer - "+str(seed)+" - "+datasetFolder.replace("\\","")+".zip",home+"\\BERandomizer - "+str(seed)+" - "+datasetFolder.replace("\\","")+".mcpack")

def cleanup(path):
    print("Cleaning up")
    deldir(path)

modules/autoupdate.py
import subprocess, sys, os
from modules import settings, dirs
currentString = "Non-Git"
branchString = "Non-Git"

def updateGit():
    '''Checks for updates if autoUpdate is True'''
    if(settings.settings['autoUpdate']==True):
        try: #put incase 'git' is not installed on computer
            global currentString
            global branchString
            per = '%'
            subprocess.run('git fetch',shell=True,stdout=subprocess.DEVNULL,stderr=subprocess.DEVNULL)
            branchCommit = subprocess.check_output('git log -n 1 --date=raw-local --pretty=format:"'+per+'h '+per+'ad" origin/master',shell=True)
            localCommit = subprocess.check_output('git log -n 1 --date=raw-local --pretty=format:"'+per+'h '+per+'ad"',shell=True)
            branchCommit = branchCommit.split()
            localCommit = localCommit.split()
            branchString = branchCommit[0].decode('utf-8')
            currentString = localCommit[0].decode('utf-8')
            if(int(branchCommit[1])>int(localCommit[1])):
                return(True)
            else:
                return(False)
        except:
            settings.setUpdate(False)
            settings.writeSettings(dirs.dataDir+"\\settings.json")
            currentString = "Git not installed"
            return(False)

def update():
    '''Handles the user-side of installing updates'''
    if(os.path.exists('.git')==True):
        print("Checking for updates...", end="",flush=True)
        if(updateGit()==True):
            print("\tUpdate found")
            update = input("Would you like to install?")
            if('y' in update.lower()):
                subprocess.run(['python','update.py'],shell=True)
                sys.exit()
            if('n' in update.lower()):
                pass
        else:
            print("\tNo updates available")
    else:
        settings.setUpdate(False)
        settings.writeSettings(dirs.dataDir+"\\settings.json")
        print("Automatic updates disabled, was not downloaded via 'git clone'.")

modules/recipes.py
import os, json, random, subprocess
from modules.debug import debug
from modules import dirs, settings
loadlist = {
    'minecraft:recipe_shaped': 'result',
    'minecraft:recipe_shapeless': 'result',
    'minecraft:recipe_furnace': 'output',
    'minecraft:recipe_brewing_mix': 'output',
    'minecraft:recipe_brewing_container': 'reagent'
}
dataload = {
    'minecraft:dye' : 'dyes',
    'minecraft:carpet' : 'colors',
    'minecraft:wool' : 'colors',
    'minecraft:bucket' : 'bucket',
    'minecraft:log' : 'log',
    'minecraft:log2' : 'log2',
    'minecraft:concrete_powder' : 'colors',
    'minecraft:banner' : 'colors',
    'minecraft:stained_glass' : 'colors',
    'minecraft:stained_glass_pane' : 'colors',
    'minecraft:red_flower' : 'flowers',
    'minecraft:fence' : 'wood',
    'minecraft:planks' : 'wood',
    'minecraft:wood' : 'wood',
    'minecraft:stained_hardened_clay' : 'colors',
    'minecraft:anvil' : 'anvil',
    'minecraft:banner_pattern' : 'banner_pattern',
    'minecraft:stone' : 'stone',
    'minecraft:boat' : 'wood',
    'minecraft:stonebrick' : 'stonebrick',
    'minecraft:cobblestone_wall' : 'wall',
    'minecraft:wooden_slab' : 'wood',
    'minecraft:emptymap' : 'map',
    'minecraft:dirt' : 'dirt',
    'minecraft:sandstone' : 'sandstone',
    'minecraft:red_sandstone' : 'sandstone',
    'minecraft:prismarine' : 'prismarine',
    'minecraft:coal' : 'coal',
    'minecraft:golden_apple' : 'gapple',
    'minecraft:stone_slab' : 'stone_slab',
    'minecraft:stone_slab2' : 'stone_slab2',
    'minecraft:stone_slab3' : 'stone_slab3',
    'minecraft:stone_slab4' : 'stone_slab4',
    'minecraft:double_stone_slab' : 'stone_slab',
    'minecraft:double_stone_slab2' : 'stone_slab2',
    'minecraft:double_stone_slab3' : 'stone_slab3',
    'minecraft:double_stone_slab4' : 'stone_slab4',
    'minecraft:quartz_block' : 'quartz',
    'minecraft:stone_button' : 'ignore',
    'minecraft:dropper' : 'ignore'
}

documentation = {
    'dyes' : {
        0 : 'ink sac',
        1 : 'red',
        2 : 'green',
        3 : 'cocoa beans',
        4 : 'lapis',
        5 : 'purple',
        6 : 'cyan',
        7 : 'light gray',
        8 : 'gray',
        9 : 'pink',
        10 : 'lime',
        11 : 'yellow',
        12 : 'light blue',
        13 : 'magenta',
        14 : 'orange',
        15 : 'bonemeal',
        16 : 'black',
        17 : 'brown',
        18 : 'blue',
        19 : 'white'
    },
    'colors' : {
        0 : 'white',
        1 : 'orange',
        2 : 'magenta',
        3 : 'light blue',
        4 : 'yellow',
        5 : 'lime',
        6 : 'pink',
        7 : 'gray',
        8 : 'light gray',
        9 : 'cyan',
        10 : 'purple',
        11 : 'blue',
        12 : 'brown',
        13 : 'green',
        14 : 'red',
        15 : 'black'
    },
    'bucket' : {
        0 : 'empty',
        1 : 'milk',
        2 : 'cod',
        3 : 'salmon',
        4 : 'tropical',
        5 : 'pufferfish',
        8 : 'water',
        10 : 'lava'
    },
    'log' : {
        0 : 'oak',
        1 : 'spruce',
        2 : 'birch',
        3 : 'jungle'
    },
    'log2' : {
        0 : 'acacia',
        1 : 'dark oak'
    },
    'wood' : {
        0 : 'oak',
        1 : 'spruce',
        2 : 'birch',
        3 : 'jungle',
        4 : 'acacia',
        5 : 'dark oak',
        8 : 'stripped oak',
        9 : 'stripped spruce',
        10 : 'stripped birch',
        11 : 'stripped jungle',
        12 : 'stripped acacia',
        13 : 'stripped dark oak'
    },
    'flowers' : {
        0 : 'poppy',
        1 : 'blue orchid',
        2 : 'allium',
        3 : 'azure bluet',
        4 : 'red tulip',
        5 : 'orange tulip',
        6 : 'white tulip',
        7 : 'pink tulip',
        8 : 'oxeye daisy',
        9 : 'cornflower',
        10 : 'lily of the valley'
    },
    'anvil' : {
        0 : 'undamaged',
        1 : 'slightly damaged',
        2 : 'very damaged'
    },
    'banner_pattern' : {
        0 : 'creeper charge',
        1 : 'skull charge',
        2 : 'flower charge',
        3 : 'thing',
        4 : 'field masoned',
        5 : 'bordure indented'
    },
    'stone' : {
        0 : 'stone',
        1 : 'granite',
        2 : 'polished granite',
        3 : 'diorite',
        4 : 'polished diorite',
        5 : 'andesite',
        6 : 'polished andesite'
    },
    'stonebrick' : {
        0 : 'stone brick',
        1 : 'mossy brick',
        2 : 'cracked brick',
        3 : 'chiseled brick'
    },
    'wall' : {
        0 : 'cobble',
        1 : 'mossycobble',
        2 : 'granite',
        3 : 'diorite',
        4 : 'andesite',
        5 : 'sandstone',
        6 : 'brick',
        7 : 'stonebrick',
        8 : 'mossybrick',
        9 : 'endbrick',
        10 : 'netherbrick',
        11 : 'prismarine',
        12 : 'redsandstone',
        13 : 'rednetherbrick'
    },
    'map' : {
        0 : 'normal',
        2 : 'locator'
    },
    'dirt' : {
        0 : 'dirt',
        1 : 'coarse'
    },
    'coal' : {
        0 : 'coal',
        1 : 'charcoal'
    },
    'prismarine' : {
        0 : 'prismarine',
        1 : 'dark prismarine',
        2 : 'prismarine bricks'
    },
    'sandstone' : {
        0 : 'normal',
        1 : 'chiseled',
        2 : 'cut',
        3 : 'smooth'
    },
    'gapple' : {
        0 : 'normal',
        1 : 'enchanted'
    },
    'stone_slab' : {
        0 : 'smooth',
        1 : 'sandstone',
        2 : 'wooden',
        3 : 'cobblestone',
        4 : 'bricks',
        5 : 'stonebrick',
        6 : 'quartz',
        7 : 'netherbrick'
    },
    'stone_slab2' : {
        0 : 'red_sandstone',
        1 : 'purpur',
        2 : 'prismarine',
        3 : 'prismarine brick',
        4 : 'dark prismarine',
        5 : 'mossy cobble',
        6 : 'smooth sandstone',
        7 : 'red netherbrick',
    },
    'stone_slab3' : {
        0 : 'end brick',
        1 : 'smooth red sandstone',
        2 : 'polished andesite',
        3 : 'andesite',
        4 : 'diorite',
        5 : 'polished_diorite',
        6 : 'granite',
        7 : 'polished granite'
    },
    'stone_slab4' : {
        0 : 'mossy brick',
        1 : 'smooth quartz',
        2 : 'stone',
        3 : 'cut sandstone',
        4 : 'cut red sandstone'
    },
    'ignore' : 'ignore',
    'quartz' : {
        0 : 'normal',
        1 : 'chiseled',
        2 : 'pillar',
        3 : 'smooth',
        5 : 'chiseled',
        6 : 'pillar',
        9 : 'chiseled',
        10 : 'pillar',
        13 : 'chiseled',
        14 : 'pillar'
    }
}

outcomes = []
scrambles = []

def blacklistcheck(blacklist,rtype):
    if(rtype in blacklist):
        return(True)
    else:
        return(False)

def loaddata(data,store,path):
    '''Loads data given to it'''
    datakeys = data.keys()
    for key in loadlist:
        itemData = None
        if(key in datakeys):
            id = data[key]['description']['identifier']
            if(blacklistcheck(settings.settings['recipeBlacklist'],key)==False):
                outcome = data[key][loadlist[key]]
                dirs.makedir(dirs.tempDir+"\\"+dirs.datasetFolder) #make temp folder
                if(isinstance(outcome,list)): #if output is list of dicts
                    items = [] #create list of items
                    for item in outcome: #add each item to list
                        if('data' in item):
                            itemData = item['data']
                            items.append(item['item']+"["+str(documentation[dataload[item['item']]][itemData])+"]")
                        else:
                            items.append(item['item'])
                    store.append(items)
                    items = ",".join(items) #turn items into string
                    itemData = None
                    document(id,str(items),path,itemData)
                elif(isinstance(outcome,dict)): #if output is one dict
                    if('data' in outcome):
                        itemData = data[key][loadlist[key]]['data']
                    store.append(outcome) #store the item
                    document(id,outcome['item'],path,itemData)
                elif(isinstance(outcome,str)):
                    store.append(outcome)
                    document(id,outcome,path,itemData)
            if(blacklistcheck(settings.settings['recipeBlacklist'],key)==True):
                print("Skipped blacklisted recipe "+id)
            debug(str(len(outcomes))+"/"+str(dirs.countDir(dirs.dataDir+"\\"+dirs.datasetFolder+"\\recipes"))+" outcomes loaded.")

def document(recipe,result,path,extradata): #used to create recipe documentation
    '''Creates recipe documentation'''
    if(extradata is None): 
        message = recipe+" = "+result+"\n"
    else:
        try:
            if(documentation[dataload[result]] is 'ignore'):
                message = recipe+" = "+result+"\n"
        except:
            pass
        try:
            message = recipe+" = "+result+"["+documentation[dataload[result]][extradata]+"]\n"
        except:
            message = recipe+" = "+result+"\n"
            print("Data error with recipe "+recipe+" and data "+str(extradata))
    with open(path,"a+") as file:
            file.write(message)

def scrambledata(data,store,path):
    '''Scrambles data given to it'''
    datakeys = data.keys()
    new_outcome = random.choice(outcomes)
    for key in loadlist:
        itemData = None
        if(key in datakeys):
            id = data[key]['description']['identifier']
            data[key][loadlist[key]] = new_outcome
            if(isinstance(new_outcome,list)): #if output is list of dicts
                items = [] #create list of items
                for item in new_outcome: #add each item to list
                    if('data' in item):
                        itemData = item['data']
                        items.append(item['item']+"["+str(documentation[dataload[item['item']]][itemData])+"]")
                    else:
                        items.append(item)
                store.append(items)
                items = ",".join(items) #turn items into string
                itemData = None
                document(id,items,path,itemData)
            elif(isinstance(new_outcome,dict)): #if output is one dict
                if('data' in new_outcome):
                        itemData = data[key][loadlist[key]]['data']
                store.append(new_outcome['item'])
                document(id,new_outcome['item'],path,itemData)
            elif(isinstance(new_outcome,str)): #if output is string
                store.append(new_outcome)
                document(id,new_outcome,path,itemData)
            outcomes.remove(new_outcome)
            debug("Scrambled: "+str(len(scrambles)))
    return data

def load():
    '''Load recipes from data'''
    #load data
    print("Loading outcomes from data... ",end="",flush=True)
    for file in os.listdir(dirs.dataDir+dirs.datasetFolder+"\\recipes"):
        with open(dirs.dataDir+dirs.datasetFolder+"\\recipes\\"+file) as json_file:
            data = json.load(json_file)
            loaddata(data,outcomes,dirs.vanilla)
    print("Loaded "+str(len(outcomes))+" outcomes.")

def scramble():
    '''Scramble recipes'''
    #scrambles data
    print("Scrambling data... ",end="",flush=True)
    for file in os.listdir(dirs.tempDir+dirs.datasetFolder+"\\recipes"):
        deletefile = False
        with open(dirs.tempDir+dirs.datasetFolder+"\\recipes\\"+file,"r") as json_file:
            data = json.load(json_file)
            datakeys = data.keys()
        for key in datakeys: #checks if file is in blacklist
            if(blacklistcheck(settings.settings['recipeBlacklist'],key)==True):
                json_file.close()
                deletefile = True
        if(deletefile==True):
            os.remove(json_file.name)
            debug("Removed "+json_file.name)
            continue
        with open(dirs.tempDir+dirs.datasetFolder+"\\recipes\\"+file,"w+") as json_file:
            json.dump(scrambledata(data,scrambles,dirs.randomized),json_file,indent=4)
    print("Scrambled "+str(len(scrambles))+" outcomes.")

def recipestart():
    '''Start randomizing recipes'''
    print("Starting recipe randomization")
    load()
    dirs.copydir(dirs.dataDir+dirs.datasetFolder+"\\recipes",dirs.tempDir+dirs.datasetFolder+"\\recipes")
    document("Seed",str(dirs.seed),dirs.randomized,None)
    scramble()